synapseclient/producer_consumer_download_threads.py

```diff
@@ -0,0 +1,282 @@
+    producer_threads = []
+    consumer_threads = []
+
+    for request in download_requests:
+        producer_threads_local = []
+        consumer_threads_local = []
+        for _ in range(num_threads):
+            producer_worker = ProducerDownloadThread(client, request, range_queue, data_queue)
+            consumer_worker = ConsumerDownloadThread(data_queue)
+            producer_threads_local.append(producer_worker)
+            consumer_threads_local.append(consumer_worker)
+
+        for producer_thread, consumer_thread in zip(producer_threads_local, consumer_threads_local):
+            producer_thread.start()
+            consumer_thread.start()
+
+        for producer_thread, consumer_thread in zip(producer_threads_local, consumer_threads_local):
+            producer_threads.append(producer_thread)
+            consumer_threads.append(consumer_thread)
+
+        global transferred
+        global t_first_write
+        global previously_transferred
+        transferred = t_first_write = previously_transferred = 0
+
+        file_name, pre_signed_url = _get_pre_signed_batch_request_json(client, request)
+        res_get = session.get(pre_signed_url, stream=True)
+        file_size = int(res_get.headers['Content-Length'])
+        pre_signed_url_chunk_generator = _get_chunk_pre_signed_url(file_size,
+                                                                   file_name,
+                                                                   pre_signed_url,
+                                                                   request.path)
+        global to_be_transferred
+        to_be_transferred = file_size
+        _create_empty_file(file_size, request.path)
+        for chunk in pre_signed_url_chunk_generator:
+            range_queue.put(chunk)
+    range_queue.join()
+    data_queue.join()
+
+
+def _get_pre_signed_batch_request_json(client, request: DownloadRequest) -> tuple:
+    """
+    Returns the file_name and pre-signed url for download as specified in request
+
+    :param client: A SynapseBaseClient
+    :param request: An individual entry in the form of a DownloadRequest
+    :return: A tuple containing the file_name and pre-signed url
+    """
+    pre_signed_url_request = {
+                                "requestedFiles": [
+                                    {
+                                        "fileHandleId": request.file_handle_id,
+                                        "associateObjectId": request.object_id,
+                                        "associateObjectType": request.object_type
+                                    }
+                                ],
+                                "includePreSignedURLs": True,
+                                "includeFileHandles": True,
+                                "includePreviewPreSignedURLs": False
+                            }
+    pre_signed_url_batch = client.restPOST(uri=SYNAPSE_URL_PATH_DOWNLOAD_GET_BATCH_PRE_SIGNED_URL,
```

Comment:
Try to replace with client method that gets pre signed urls

Reply:
Can do this (will do for now), however it is accessing a protected member of the client class which is not good style/discouraged. Since we know that accessing this protected member won't mutate the state of the client we will use it in this patch, in v2 no protected members should be accessed outside of the class they are declared.